src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs

```diff
@@ -45,5 +50,289 @@ protected override void InsertItem(int index, IDeltaSetItem item)
 
             base.InsertItem(index, item);
         }
+
+        /// <summary>
+        /// Patch for DeltaSet, a collection for Delta<typeparamref name="TStructuralType"/>
+        /// </summary>
+        /// <param name="originalCollection">Original collection of the Type which needs to be updated</param>
+        /// /// <returns>DeltaSet response</returns>
+        public DeltaSet<TStructuralType> Patch(ICollection<TStructuralType> originalCollection)
+        {
+            ODataAPIHandler<TStructuralType> apiHandler = new DefaultODataAPIHandler<TStructuralType>(originalCollection);
+
+            return CopyChangedValues(apiHandler);
+        }
+
+
+        /// <summary>
+        /// Patch for DeltaSet, a collection for Delta<typeparamref name="TStructuralType"/>
+        /// </summary>     
+        /// <returns>DeltaSet response</returns>
+        public DeltaSet<TStructuralType> Patch(ODataAPIHandler<TStructuralType> apiHandlerOfT, ODataAPIHandlerFactory apiHandlerFactory)
+        {         
+            Debug.Assert(apiHandlerOfT != null);
+
+            return CopyChangedValues(apiHandlerOfT, apiHandlerFactory);
+        }
+
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes")]
+        internal DeltaSet<TStructuralType> CopyChangedValues(IODataAPIHandler apiHandler, ODataAPIHandlerFactory apiHandlerFactory = null)
+        {
+            //Here we are getting the keys and using the keys to find the original object 
+            //to patch from the list of collection
+
+            ODataAPIHandler<TStructuralType> apiHandlerOfT = apiHandler as ODataAPIHandler<TStructuralType>;
+
+            Debug.Assert(apiHandlerOfT != null);
+
+            DeltaSet<TStructuralType> deltaSet = CreateDeltaSet();
+
+            foreach (Delta<TStructuralType> changedObj in Items)
+            {
+                DataModificationOperationKind operation = DataModificationOperationKind.Update;
+
+                //Get filtered item based on keys
+                TStructuralType original = null;
+                string errorMessage = string.Empty;
+                string getErrorMessage = string.Empty;
+
+                Dictionary<string, object> keyValues = new Dictionary<string, object>();
+
+                foreach (string key in _keys)
+                {
+                    object value;
+
+                    if (changedObj.TryGetPropertyValue(key, out value))                   
+                    {
+                        keyValues.Add(key, value);
+                    }
+                }
+
+                try
+                {
+                    ODataAPIResponseStatus ODataAPIResponseStatus = apiHandlerOfT.TryGet(changedObj.ODataPath.GetKeys(), out original, out getErrorMessage);
```

Comment:
```suggestion
                    ODataAPIResponseStatus ODataAPIResponseStatus = apiHandlerOfT.TryGet(keyValues, out original, out getErrorMessage);
```

Reply:
When we are using `odataId`, `keyValues` has the wrong values since at this point we don't know the original object. However we can get the correct keys from the `ODataPath`